### PR TITLE
@shopify/theme-predictive-search

### PR DESCRIPTION
**Pull request by antoinegrant**
### Configuration options 

<table>
  <tr>
    <th width="30%">Type of search</th>
    <th>Options</th>
  </tr>
  <tr>
    <td><code>resources</code></td>
    <td>
      <table>
        <tr>
          <th width="30%">Option</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
        <tr>
          <td><code>types</code> (required)</td>
          <td>Array</td>
          <td>Specifies the type of results requested. Currently supports <code>product</code> only.</td>
        </tr>
        <tr>
          <td><code>fuzzy</code> (optional)</td>
          <td>Boolean</td>
          <td>Enables typo tolerance when searching. A typo will be tolerated after the fourth character. Default: <code>true</code>.</td>
        </tr>
        <tr>
          <td><code>unavailable_products</code> (optional)</td>
          <td>String</td>
          <td>Specifies whether to display <code>unavailable_products</code> results. The three possible options are <code>show</code>, <code>hide</code>, and <code>bury</code>. Burying out of stock products pushes them below other matching results.  Default: <code>show</code>.</td>
        </tr>
        <tr>
          <td><code>limit</code> (optional)</td>
          <td>Integer</td>
          <td>Limits the number of results returned. Default: <code>10</code>. Min:  <code>1</code>. Max: <code>10</code>.</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

See the [help docs for Predictive Search](https://help.shopify.com/en/themes/development/predictive-search) for more information.

---

### Getting started

```javascript
import PredictiveSearch from "@shopify/theme-predictive-search";

var predictiveSearch = new PredictiveSearch({
  resources: {
    fuzzy: true,
    types: [PredictiveSearch.TYPES.PRODUCT],
    limit: 4,
    unavailable_products: "bury"
  }
});

predictiveSearch.on("success", function(json) {
  // See "Success Response" section of this document
});

predictiveSearch.on("error", function(error) {
  // See "HTTP status `3xx-4xx` Response" section of this document
});

predictiveSearch.query("The Calling");
```

---

### Success Response

```json
// JSON Output
{
  "resources": {
    "results": {
      "products": [
        {
          "id": 1,
          "title": "The Calling",
          "body": "<p>The Calling</p>",
          "handle": "calling",
          "image": "https://cdn.shopify.com/...",
          "type": "Bikes",
          "tags": ["bike", "dolphin"],
          "url": "/products/calling?variant_id=1",
          "price": "3099",
          "available": true,
          "variants": [
            {
              "title": "Large / Angry Dolphin",
              "url": "https://www.evil-bikes.com/products/calling",
              "image": "https://cdn.shopify.com/...",
              "price": "3099",
              "compare_at_price": "4099",
              "available": true,
            }
          ]
        }
      ]
    }
  }
}
```

### HTTP status `3xx-4xx` Response

```js
predictiveSearch.on("error", function(error) {
  console.error(error.status); // Ex. output: 429
  console.error(error.name); // Ex. output: Throttled
  console.error(error.message); // Ex. output: Too Many Requests

  // When the request response HTTP status is 429
  console.error(error.retryAfter); // Ex. output: 100
});
```

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (antoinegrant) [inline .eslintrc]: This change is to match patterns found in other repos at Shopify.

**Comment by antoinegrant** [comment]
> Do you think the api utilities can be break out into different package?

In order to break out these into a utilities package, we would need to figure out cross package dependencies in the case where the packages are not consumed in a build process but rather copied to a theme.

**Comment by wizardlyhel** [comment]
Are you talking about the compiled script that we provide as a fallback for those aren't using any package manager?

If you are, I believe that the compile script should contain the included api utilities so that the script is complete and not separated. So that the usage of this library isn't like .. include these 5 scripts, in this order, to get predictive search working.

We do discourage the use of script link as a way to use the library due to performance reasons:
* Potential of duplicated includes of the same utility library
* Extra commonJs wrapper for each library script includes

However, we shouldn't abandon the majority of storefronts that isn't on some kinda package manager either and this is what it comes down to: An impact that we decided to accept, for now.

File: packages/theme-predictive-search/src/theme-predictive-search.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import { validateConfig, validateQuery } from "./validate";
+import request, { parseResponseStatus, parseResponse } from "./request";
+import { debounce, objectToQueryParams, Dispatcher, Cache } from "./utilities";
+
+export default function PredictiveSearch(config) {
+  try {
+    validateConfig(config);
+  } catch (error) {
```
Comment (t-kelly) [inline packages/theme-predictive-search/src/theme-predictive-search.js]: Is this catch needed? You're just rethrowing the error?

File: packages/theme-predictive-search/src/theme-predictive-search.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+PredictiveSearch.TYPES = {
+  PRODUCT: "product"
+};
+
+PredictiveSearch.prototype.DEBOUNCE_RATE = 10;
+
+PredictiveSearch.prototype.query = function query(query) {
+  var self = this;
```
Comment (t-kelly) [inline packages/theme-predictive-search/src/theme-predictive-search.js]: Do we really need `self`?

File: packages/theme-predictive-search/src/theme-predictive-search.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  try {
+    validateQuery(query);
+  } catch (error) {
+    this.dispatcher.dispatch("error", error);
+    return;
+  }
+
+  if (query === "") {
+    return self;
```
Comment (t-kelly) [inline packages/theme-predictive-search/src/theme-predictive-search.js]: `return self;` === `return this;` === `return;` -- `self` is redundant?

File: packages/theme-predictive-search/src/utilities/Cache.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+export default function Cache(config) {
+  this._store = {};
```
Comment (t-kelly) [inline packages/theme-predictive-search/src/utilities/Cache.js]: What is the expected size of this store? Is it a threat to memory?

File: packages/theme-predictive-search/__tests__/theme-predictive-search.test.js [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+};
+
+describe("Search()", () => {
+  it("should be able to create an instance", () => {
+    const search = new PredictiveSearch(defaultConfig);
+    expect(search).toBeInstanceOf(PredictiveSearch);
+  });
+
+  describe("throws", () => {
+    it("when config object is not set", () => {
```
Comment (huguestennier) [inline packages/theme-predictive-search/__tests__/theme-predictive-search.test.js]: Wonder if it's possible to have good defaults so we wouldn't need this. What do you think?

File: packages/theme-predictive-search/src/theme-predictive-search.js (hunk 2 above)
Comment (antoinegrant) [reply packages/theme-predictive-search/src/theme-predictive-search.js]: Initially I was planning on having more validation around `config`, but it doesn't scale very well as we add more features to the API. I'll simply let the API return an errors if the config is not valide.

File: packages/theme-predictive-search/src/utilities/Cache.js (hunk 5 above)
Comment (antoinegrant) [reply packages/theme-predictive-search/src/utilities/Cache.js]: I did want to limit the memory used by this cached results. Do you have some thoughts around that pattern?

File: packages/theme-predictive-search/README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
+## Shopify AJAX Search API
+
+### Getting started
+
+```javascript
+import PredictiveSearch from "@shopify/theme-predictive-search";
+
+var predictiveSearch = new PredictiveSearch({
+  search_as_you_type: {
+    fuzzy: true,
```
[the remark this answers is not in the document]
Comment (antoinegrant) [reply packages/theme-predictive-search/README.md]: I should add way more documentation to the README. I will make sure we do clarify what fuzzy means.

File: packages/theme-predictive-search/README.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    types: [PredictiveSearch.TYPES.PRODUCT]
```
[the remark this answers is not in the document]
Comment (antoinegrant) [reply packages/theme-predictive-search/README.md]: Again, the README will have all the info and links to the help docs on https://help.shopify.com

**Comment by antoinegrant** [comment]
> Are you talking about the compiled script that we provide as a fallback for those aren't using any package manager?
> 
> If you are, I believe that the compile script should contain the included api utilities so that the script is complete and not separated. So that the usage of this library isn't like .. include these 5 scripts, in this order, to get predictive search working.
> 
> We do discourage the use of script link as a way to use the library due to performance reasons:
> 
> * Potential of duplicated includes of the same utility library
> * Extra commonJs wrapper for each library script includes
> 
> However, we shouldn't abandon the majority of storefronts that isn't on some kinda package manager either and this is what it comes down to: An impact that we decided to accept, for now.

### Bundling dependencies

Pros:
- Bundles the version of the util that the package expects.
- Self containing bundle makes it easy to include in a theme. (Especially true for merchants)

Cons:
- Duplicates utilities in every package that has the same dependencies.

### Not bundling dependencies

Pros:
- Does not duplicate code

Cons:
- We can't make sure that we get the version of the dependency that a package is expecting.
- You need to know to include the dependencies in the theme in the right order. (This could be explained in the README)

---

We do currently have one package that has a dependency on another theme-scripts: [`@shopify/theme-product-form` has a dependency on `@shopify/theme-product`](https://github.com/Shopify/theme-scripts/blob/master/packages/theme-product-form/theme-product-form.js#L2). In this case [we do not bundle the dependency](https://github.com/Shopify/theme-scripts/blob/master/packages/theme-product-form/rollup.config.js#L17).

CC @t-kelly @wizardlyhel @huguestennier

**Comment by wizardlyhel** [comment]
@antoinegrant I didn't realize that `theme-product-form` is dependent on `theme-product` to exist in global. This isn't documented anywhere in README either. I was under the assumption that it was all bundled together.

Let's leave it what you had - all sub packages bundle into one script.

We definitely need to change `theme-product-form` package dependency to either include it as part of compiled script or update in README (This is separate PR).

We can think about breaking these sub packages into its own packages when the time comes.

**Comment by t-kelly** [comment]
I agree with Helen 👍

File: README.md [hunk 9]
```diff
@@ -9,17 +9,18 @@ Theme scripts are uncoupled from a particular UI. Typically, they are used along
...
+| ----------------------- | ---------------------------------------------------- | -------------------------------------------------------------------------------- |
+| theme-a11y              | [README](packages/theme-a11y/README.md)              | ![npm version](https://badge.fury.io/js/%40shopify%2Ftheme-a11y.svg)             |
+| theme-addresses         | [README](packages/theme-addresses/README.md)         | ![npm version](https://badge.fury.io/js/%40shopify%2Ftheme-addresses.svg)        |
+| theme-cart              | [README](packages/theme-cart/README.md)              | ![npm version](https://badge.fury.io/js/%40shopify%2Ftheme-cart.svg)             |
+| theme-currency          | [README](packages/theme-currency/README.md)          | ![npm version](https://badge.fury.io/js/%40shopify%2Ftheme-currency.svg)         |
+| theme-images            | [README](packages/theme-images/README.md)            | ![npm version](https://badge.fury.io/js/%40shopify%2Ftheme-images.svg)           |
+| theme-predictive-search | [README](packages/theme-predictive-search/README.md) | ![npm version](https://badge.fury.io/js/%40shopify%2Ftheme-predictive-searc.svg) |
```
Comment (chrisberthe) [inline README.md]:
```suggestion
| theme-predictive-search | [README](packages/theme-predictive-search/README.md) | ![npm version](https://badge.fury.io/js/%40shopify%2Ftheme-predictive-search.svg) |
```

File: packages/theme-predictive-search/README.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  "search_as_you_type": {
+    "results": {
+      "products": [
+        {
+          "title": "The Calling",
+          "body": "<p>The Calling</p>",
+          "handle": "calling",
+          "image": "https://cdn.shopify.com/...",
+          "url": "/products/calling?variant_id=1",
+          "price": "3099.00",
```
Comment (huguestennier) [inline packages/theme-predictive-search/README.md]:
```suggestion
          "price": "3099",
```

File: packages/theme-predictive-search/README.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+          "variants": [
+            {
+              "title": "Large / Angry Dolphin",
+              "url": "https://www.evil-bikes.com/products/calling",
+              "image": "https://cdn.shopify.com/...",
+              "price": "3099.00",
```
[the remark this answers is not in the document]
Comment (huguestennier) [reply packages/theme-predictive-search/README.md]: This is marked as resolved, but it still says `3099.00`. Should be `3099`

```suggestion
              "price": "3099",
```

File: packages/theme-predictive-search/README.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+              "compare_at_price": "4099.00"
```
Comment (huguestennier) [inline packages/theme-predictive-search/README.md]:
```suggestion
              "compare_at_price": "4099"
```

File: packages/theme-predictive-search/__fixtures__/search_as_you_type_the_following.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  "search_as_you_type": {
+    "results": {
+      "products": [
+        {
+          "title": "The Following",
+          "body": "<p>The Following</p>",
+          "handle": "following",
+          "image": "https://cdn.shopify.com/...",
+          "url": "/products/following?variant_id=1",
+          "price": "3099.00",
```
Comment (huguestennier) [inline packages/theme-predictive-search/__fixtures__/search_as_you_type_the_following.json]:
```suggestion
          "price": "3099",
```

File: packages/theme-predictive-search/__fixtures__/search_as_you_type_the_following.json [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+          "variants": [
+            {
+              "title": "Large / Angry Dolphin",
+              "url": "https://www.evil-bikes.com/products/following",
+              "image": "https://cdn.shopify.com/...",
+              "price": "3099.00",
```
Comment (huguestennier) [inline packages/theme-predictive-search/__fixtures__/search_as_you_type_the_following.json]:
```suggestion
              "price": "3099",
```

File: packages/theme-predictive-search/__fixtures__/search_as_you_type_the_following.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+              "compare_at_price": "4099.00"
```
Comment (huguestennier) [inline packages/theme-predictive-search/__fixtures__/search_as_you_type_the_following.json]:
```suggestion
              "compare_at_price": "4099"
```

File: packages/theme-predictive-search/README.md [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+              "price": "3099",
+              "compare_at_price": "4099"
+            }
+          ]
+        }
+      ]
+    }
+  }
+}
+```
```
Comment (BrendaSegalShopify) [inline packages/theme-predictive-search/README.md]: Not sure if we also want to add `id`, `available`, `tags`, etc here so:

```
{
  "search_as_you_type": {
    "results": {
      "products": [
This conversation was marked as resolved by antoinegrant
        {
          "id": 5,
          "title": "The Calling",
          "body": "<p>The Calling</p>",
          "handle": "calling",
          "image": "https://cdn.shopify.com/...",
          "vendor": "Dolphin Bikes",
          "type": "Bikes",
          "tags": ["bike", "dolphin"],
          "url": "/products/calling?variant_id=1",
          "price": "3099",
          "available": true,
          "variants": [
            {
              "title": "Large / Angry Dolphin",
              "url": "https://www.evil-bikes.com/products/calling",
              "image": "https://cdn.shopify.com/...",
              "price": "3099",
              "compare_at_price": "4099",
              "available": true
            }
          ]
        }
      ]
    }
  }
}
```

File: packages/theme-predictive-search/src/request.js [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+          onError(throttledError);
+        } catch (error) {
+          onError(error);
+        }
+
+        return;
+      }
+
+      if (xhr.status) {
```
Comment (BrendaSegalShopify) [inline packages/theme-predictive-search/src/request.js]: Not sure if we need the `if` here => would we ever have a point when `xhr.status` is not truthy?
If we do, then what happens in that case?
I think this is fine, I'm just not 100% sure.

File: packages/theme-predictive-search/README.md (hunk 16 above)
Comment (BrendaSegalShopify) [reply packages/theme-predictive-search/README.md]: Also, one we might end up adding `product.compare_at_min` and `product.compare_at_max`.  I'm going to create an issue about this, but the above are the fields that exist currently.

Comment (BrendaSegalShopify) [reply packages/theme-predictive-search/README.md]: Relevant issue related to `compare_at_price_min`/`compare_at_price_max`: https://github.com/Shopify/shopify/issues/200166

File: packages/theme-predictive-search/README.md [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+        {
+          "id": 1,
+          "title": "The Calling",
+          "body": "<p>The Calling</p>",
+          "handle": "calling",
+          "image": "https://cdn.shopify.com/...",
+          "type": "Bikes",
+          "tags": ["bike", "dolphin"],
+          "url": "/products/calling?variant_id=1",
+          "price": "3099",
```
Comment (BrendaSegalShopify) [inline packages/theme-predictive-search/README.md]: Should we also add the following pricing fields within `product`?
```
compare_at_price_max
compare_at_price_min
price_max
price_min
```

**Pull request merged**